### PR TITLE
Support new `labelRenderer` and `valueRenderer` properties

### PR DESCRIPTION
**Pull request by bvaughn**
In order to support bvaughn/redux-devtools-filterable-log-monitor/issues/8, I need finer-grained control over what is rendered for node keys and values. So I set out to add 2 new properties: `labelRenderer` and `valueRenderer`.

Along the way I also noticed a _lot_ of repetition between classes (`JSONBooleanNode`, `JSONDateNode`, `JSONFunctionNode`, `JSONNullNode`, `JSONNumberNode`, and `JSONUndefinedNode`- and even `JSONArrayNode`, `JSONIterableNode`, and `JSONObjectNode`). So I did some refactoring to tidy things up a bit. I hope I didn't take too much liberty by doing this.

I'd love to see this feature merged and I'd be willing to make any changes you feel necessary. Please feel free to reach out to me and we can chat. :)

### CODE REVIEW

**Comment by bvaughn** [comment]
Temporarily closing this PR while I investigate a performance issue. Will re-open once I've identified it.

**Comment by chibicode** [comment]
@bvaughn thanks a lot! Keep me updated.

**Comment by bvaughn** [comment]
Happy to help. Will do! I noticed a performance decrease in the branch of bvaughn/redux-devtools-filterable-log-monitor that integrated with this update and I just wanted to rule out the change _here_ before submitting. I'll keep you posted. :)

**Comment by bvaughn** [comment]
I identified the performance issue. My branch accidentally introduced a regression- all nodes were expanded by default instead of just the root. Easy enough to fix. :)

File: README.md [hunk 1]
```diff
@@ -117,6 +117,19 @@ Then the preview of child elements now look like this:
...
 ![](http://cl.ly/image/1J1a0b0T0K3c/screenshot%202015-10-07%20at%203.44.31%20PM.png)
 
+#### Customize Rendering
+
+You can pass the following properties to customize rendered labels and values:
+
+```js
+<JSONTree
+    labelRenderer={raw => <strong>raw</strong>}
+    valueRenderer={raw => <em>raw</em>}
```
Comment (chibicode) [inline README.md]: Should be `<strong>{ raw }</strong>` (same for `em`) instead

**Comment by chibicode** [comment]
Added one comment but this looks good. Thank you so much for cleaning it up! (and say hi to Treasure Data folks)

I tried to run `npm start` under `examples` directory and it failed. I'm guessing it's because this PR is dependent on one of React 0.14's features, and `react-json-tree`'s `examples` is using React 0.13. Do you think that's the case?

I'm fine w/ dropping the support for 0.13.

I'd also add this to `examples/src/App.js`:

```js
<h3>More Fine Grained Rendering</h3>
<p>Pass <code>labelRenderer</code> or <code>valueRenderer</code>.</p>
<div>
  <JSONTree data={ data }
            labelRenderer={raw => <strong>{ raw }</strong>}
            valueRenderer={raw => <em>{ raw }</em>} />
</div>
```

File: README.md (hunk 1 above)
Comment (bvaughn) [inline README.md]: Whoops! Nice catch. I need to pay more attention to my code examples. :)

**Comment by bvaughn** [comment]
Not a problem! Thanks for the useful library. :)

Hm. I'm unsure about the issue you pointed out. I don't specifically remember using _any_ React 0.14.x features. That being said, updating the example package to React 0.14.x did resolve the problem- so if you don't mind me bumping that dependency with this PR- that would certainly be the most straight forward solution.

I updated the example App with the blurb you suggested. Thanks for that.

**Comment by chibicode** [comment]
@bvaughn sounds good, thank you!

**Comment by bvaughn** [comment]
You're quite welcome! :)

**Comment by chibicode** [comment]
@bvaughn published 0.4.0.

**Comment by bvaughn** [comment]
Thanks so much for the speedy update!

**Pull request merged**